clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
      * @param cluster The current cluster metadata
      */
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```

Comment:
The only thing I'm puzzling over is the impact of this check. Currently, KafkaConsumer has the following snippet in poll():
```java
fetcher.initFetches(cluster);
client.poll(timeout);
return fetcher.fetchedRecords();
```
I'm wondering in particular about that `client.poll(timeout)`. If we haven't actually initiated any new fetches, then there wouldn't be any expected IO and we'd block for the timeout duration unnecessarily. You could instead change it to something like this:
```java
Map records = fetcher.fetchedRecords();
if (!records.isEmpty())
  return records;
fetcher.initFetches(cluster);
client.poll(timeout);
return fetcher.fetchedRecords();
```
Then we could remove this check and we wouldn't block unless there actually were pending fetches. What do you think?